### PR TITLE
Reading a HDF5 created in py2

### LINKED ISSUE #24925: reading of old pandas dataframe (created in python 2) failed with 0.23.4
Hi,

Firstly I have to apologize, that my description will be very vague.

I have a problem with one of my dataframe that was created earlier with python 2 and older version of pandas (unfortunately I do not know what version). Now I cannot open it in python 3 and pandas 0.23.4 (loading in python 3 with pandas 0.22.0 works fine).

For reading, I am using:

```python
hdf = pd.HDFStore(src_filename, mode=”r”)
data_frame = hdf.select(src_tablename)
```
My stack trace in pandas 0.23.4 is:

```
Traceback (most recent call last):
    data_frame = hdf.select(src_tablename)
  File "/home/rbenes/virtual_envs/iface_venv36_new_pkgs/lib/python3.6/site-packages/pandas/io/pytables.py", line 743, in select
    return it.get_result()
  File "/home/rbenes/virtual_envs/iface_venv36_new_pkgs/lib/python3.6/site-packages/pandas/io/pytables.py", line 1485, in get_result
    results = self.func(self.start, self.stop, where)
  File "/home/rbenes/virtual_envs/iface_venv36_new_pkgs/lib/python3.6/site-packages/pandas/io/pytables.py", line 734, in func
    columns=columns)
  File "/home/rbenes/virtual_envs/iface_venv36_new_pkgs/lib/python3.6/site-packages/pandas/io/pytables.py", line 4182, in read
    if not self.read_axes(where=where, **kwargs):
  File "/home/rbenes/virtual_envs/iface_venv36_new_pkgs/lib/python3.6/site-packages/pandas/io/pytables.py", line 3385, in read_axes
    errors=self.errors)
  File "/home/rbenes/virtual_envs/iface_venv36_new_pkgs/lib/python3.6/site-packages/pandas/io/pytables.py", line 2195, in convert
    self.data, nan_rep=nan_rep, encoding=encoding, errors=errors)
  File "/home/rbenes/virtual_envs/iface_venv36_new_pkgs/lib/python3.6/site-packages/pandas/io/pytables.py", line 4658, in _unconvert_string_array
    data = libwriters.string_array_replace_from_nan_rep(data, nan_rep)
  File "pandas/_libs/writers.pyx", line 158, in pandas._libs.writers.string_array_replace_from_nan_rep
ValueError: Buffer dtype mismatch, expected 'Python object' but got 'double'
```

This stack trace led me to this pull request: https://github.com/pandas-dev/pandas/pull/24510

If I list it e.g. with h5ls it looks fine (it is loaded and content looks fine).

Unfortunately, I cannot share the dataframe, because it is private and I cannot reproduce process of the creation with older versions any more :-(. So I am not able to deliver that unreable dataframe.

I debuged pandas and found, that this patch helped me. 

```
diff --git a/pandas/io/pytables.py b/pandas/io/pytables.py
index 4e103482f..2ab6ddb5b 100644
--- a/pandas/io/pytables.py
+++ b/pandas/io/pytables.py
@@ -3288,7 +3288,7 @@ class Table(Fixed):
         self.nan_rep = getattr(self.attrs, 'nan_rep', None)
         self.encoding = _ensure_encoding(
             getattr(self.attrs, 'encoding', None))
-        self.errors = getattr(self.attrs, 'errors', 'strict')
+        self.errors = _ensure_decoded(getattr(self.attrs, 'errors', 'strict'))
         self.levels = getattr(
             self.attrs, 'levels', None) or []
         self.index_axes = [
```
Can anyone advice me, if such a fix is fine and if yes, can I send it as pull request without any reproducer? 

Thank you.

### PR DESCRIPTION
**Pull request by rbenes**
- [ x] closes #24925
- [ x] tests added / passed
- [ x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.24.1.rst [hunk 1]
```diff
@@ -50,7 +50,7 @@ Bug Fixes
 
 **I/O**
 
--
+- Bug in :class:`HDFStore` that caused it to raise ``ValueError`` when reading a Dataframe in Python 3 from dataframe written in Python 2 (:issue:`24925`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.1.rst]: I think we'll have you move this 0.24.2.

File: pandas/tests/io/test_pytables.py [hunk 2]
```diff
@@ -4552,6 +4552,20 @@ def test_legacy_table_fixed_format_read_py2(self, datapath):
...
 
+    def test_legacy_table_read_py2(self, datapath):
+        # issue: 24925
+        # legacy table written in Python 2
+        with ensure_clean_store(
+                datapath('io', 'data', 'legacy_hdf',
+                         'legacy_table_py2.h5'),
+                mode='r') as store:
+            result = store.select('table')
+            expected = pd.DataFrame({
```
Comment (TomAugspurger) [inline pandas/tests/io/test_pytables.py]: Construct expected and do the assertion outside of the `ensure_clean_store` block.

File: doc/source/whatsnew/v0.24.1.rst (hunk 1 above)
Comment (rbenes) [reply doc/source/whatsnew/v0.24.1.rst]: The v0.24.2.rst document is not created (https://github.com/pandas-dev/pandas/tree/master/doc/source/whatsnew), so I added it to this ```v0.24.1.rst```. Moreover, I think that last release is v0.24.0. Am I right?

File: pandas/tests/io/test_pytables.py (hunk 2 above)
Comment (rbenes) [reply pandas/tests/io/test_pytables.py]: changed in https://github.com/pandas-dev/pandas/pull/25058/commits/0fa7102f96fde9b2f70a2cddb070ff973963d940

File: doc/source/whatsnew/v0.24.1.rst (hunk 1 above)
Comment (jreback) [inline doc/source/whatsnew/v0.24.1.rst]: can you re-word to something like

```
Bug in reading a HDF5 table-format ``DataFrame`` created in Python 2, in Python 3
```

Comment (jreback) [reply doc/source/whatsnew/v0.24.1.rst]: am ok with 0.24.1 if you can make this change shortly.

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by TomAugspurger — APPROVED** [review]
LGTM.

Reminder, we want to wait to merge this till 0.24.1 is tagged.

Or we can merge to master, but we have to remember to *not* merge the backport PR. Probably easiest to wait to merge to master.

File: doc/source/whatsnew/v0.24.1.rst (hunk 1 above)
Comment (rbenes) [reply doc/source/whatsnew/v0.24.1.rst]: doc moved to 0.24.2, reformated to your version. Thanks.

**Review by jreback — APPROVED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.2.rst [hunk 3]
```diff
@@ -51,7 +51,7 @@ Bug Fixes
 
 **I/O**
 
--
+- Bug in reading a HDF5 table-format ``DataFrame`` created in Python 2, in Python 3
```
[the remark this answers is not in the document]
Comment (rbenes) [reply doc/source/whatsnew/v0.24.2.rst]: fixed in https://github.com/pandas-dev/pandas/pull/25058/commits/12043014ba2edb80089a88d345cb49805cc9ab0a

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @rbenes

**Pull request merged**